bpo-24416: Return a IsoCalendarDate from date.isocalendar()

corona10 opened the pull request:
<!-- issue-number: [bpo-24416](https://bugs.python.org/issue24416) -->
https://bugs.python.org/issue24416
<!-- /issue-number -->

corona10 committed:
bpo-24416: Return a namedtuple from date.isocalendar()

taleinat commented on Doc/library/datetime.rst:
```diff
@@ -590,7 +590,7 @@ Instance methods:
 .. method:: date.isocalendar()
 
-   Return a 3-tuple, (ISO year, ISO week number, ISO weekday).
+   Return a 3-namedtuple, (ISO year, ISO week number, ISO weekday).
```
I don't believe "3-namedtuple" is in customary use, used elsewhere in our docs, nor clear enough in this context. Just use "named tuple".

Regardless, we'll need to describe the new `IsoCalendarDate` class explicitly (somewhere). So perhaps reference that instead.

taleinat commented on Doc/whatsnew/3.9.rst:
```diff
@@ -109,6 +109,14 @@ New Modules
+--------
+The :meth:`~datetime.date.isocalendar()` of :class:`datetime.date`
+and :meth:`~datetime.datetime.isocalendar()` of :class:`datetime.datetime`
+methods now returns a :func:`~collections.namedtuple` instead of a :class:`tuple`.
```
Both here and in the NEWS entry:

```suggestion
methods now return a :term:`named tuple` instead of a :class:`tuple`.
```

taleinat commented on Lib/datetime.py:
```diff
@@ -1091,7 +1094,7 @@ def isoweekday(self):
 
     def isocalendar(self):
-        """Return a 3-tuple containing ISO year, week number, and weekday.
+        """Return a 3-named tuple containing ISO year, week number, and weekday.
```
See above comment.

```suggestion
        """Return a named tuple containing ISO year, week number, and weekday.
```

taleinat commented on Lib/datetime.py:
```diff
@@ -1555,6 +1558,9 @@ def __reduce__(self):
 time.max = time(23, 59, 59, 999999)
 time.resolution = timedelta(microseconds=1)
 
+IsoCalendarDate = namedtuple('IsoCalendarDate', 'year week weekday')
```
Why is this here in the middle of the file, after its first use? This seems a rather strange location. Perhaps nearer the beginning of the file?

taleinat commented on Lib/test/datetimetester.py:
```diff
@@ -1363,6 +1364,13 @@ def test_isocalendar(self):
+        self.assertEqual(t.year, 2008)
+        self.assertEqual(t.week, 49)
+        self.assertEqual(t.weekday, 3)
+
```
Great job adding this test!

I suggest also checking equality with:
1. an explicitly created `IsoCalendarDate` instance
2. an equivalent 3-tuple (for backwards compatibility)

taleinat commented on Doc/whatsnew/3.9.rst:
```diff
@@ -109,6 +109,14 @@ New Modules
 
+datetime
+--------
+The :meth:`~datetime.date.isocalendar()` of :class:`datetime.date`
```
This wording seems overly verbose. I suggest:

```ReST
The ``isocalendar()`` methods of :class:`datetime.date` and :class:`datetime.datetime` now return ...
```

taleinat reviewed (changes requested):
In addition to the in-line comments, note also that the new `IsoCalendarDate` class needs its own (small) section in the docs.

ZackerySpytz commented on Modules/_datetimemodule.c:
```diff
@@ -3244,7 +3267,16 @@ date_isocalendar(PyDateTime_Date *self, PyObject *Py_UNUSED(ignored))
+    PyStructSequence_SET_ITEM(v, 2, PyLong_FromLong(day + 1));
+
+    if (PyErr_Occurred()) {
+        Py_XDECREF(v);
```
```suggestion
        Py_DECREF(v);
```

ZackerySpytz commented on Modules/_datetimemodule.c:
```diff
@@ -3234,6 +3254,9 @@ date_isocalendar(PyDateTime_Date *self, PyObject *Py_UNUSED(ignored))
     int  week;
     int  day;
 
+    PyObject* v = PyStructSequence_New(&StructIsoCalendarDateType);
```
```suggestion
    PyObject *v = PyStructSequence_New(&StructIsoCalendarDateType);
```
Also, I think this `PyStructSequence_New()` call should be moved down (so that it is closer to the `PyStructSequence_SET_ITEM()` calls) .

ZackerySpytz commented on Modules/_datetimemodule.c:
```diff
@@ -3234,6 +3254,9 @@ date_isocalendar(PyDateTime_Date *self, PyObject *Py_UNUSED(ignored))
     int  day;
 
+    PyObject* v = PyStructSequence_New(&StructIsoCalendarDateType);
+    if (!v) return NULL;
```
The `return` should not be on the same line as the `if`.
```c
if (v == NULL) {
    return NULL;
}
```
Please see the [PEP 7](https://www.python.org/dev/peps/pep-0007/) style guide.

corona10 committed:
bpo-24416: Apply code review

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -122,6 +122,14 @@ Available Types
 
    .. versionadded:: 3.2
 
+.. class:: IsoCalendarDate
```
Perhaps @taleinat will disagree with me, but if we do make this switch (and as I've said in the BPO thread, I'm not in favor of it), I consider the name and details of the return type class to be a bit of an implementation detail.

Practically speaking we would need to document that it is a `namedtuple` of some sort and what its attributes are, but ideally this would be a purely private class. I would never want to see code like `isinstance(x, datetime.IsoCalendarDate)` or `x = IsoCalendarDate(2019, 1, 1)`.

As such, if you document it anywhere it should probably not be referred to by name and just document its properties in the `isocalendar` documentation.

pganssle commented on Lib/test/datetimetester.py:
```diff
@@ -1363,6 +1364,14 @@ def test_isocalendar(self):
             d = self.theclass(2010, 1, 4+i)
             self.assertEqual(d.isocalendar(), (2010, 1, i+1))
 
+    def test_isocalendar_named(self):
```
How about refactoring this test and `test_isocalendar` together into a single parametrized test:

```python
def test_isocalendar(self):
    # Check examples from
    # https://www.phys.uu.nl/~vgent/calendar/isocalendar.htm
    week_mondays = [
        ((2003, 12, 22), (2003, 52, 1)),
        ((2003, 12, 29), (2004, 1, 1)),
        ((2004, 1, 5), (2004, 2, 1)),
        ((2009, 12, 21), (2009, 52, 1)),
        ((2009, 12, 28), (2009, 53, 1)),
        ((2010, 1, 4), (2010, 1, 1)),
    ]

    test_cases = []
    for cal_date, iso_date in week_mondays:
        base_date = self.theclass(*cal_date)

        # Adds one test case for every day of the specified weeks
        for i in range(7):
            new_date = base_date + timedelta(i)
            new_iso = iso_date[0:2] + (iso_date[2] + 1,)

            test_cases.append((new_date, new_iso))

    for d, exp_iso in test_cases:
        with self.subTest(d=d, comparison="tuple"):
            self.assertEqual(d.isocalendar(), exp_iso)

        # Check that the tuple contents are accessible by field name
        with self.subTest(d=d, comparison="fields"):
            iso_tuple = d.isocalendar()
            self.assertEqual((iso_tuple.year, iso_tuple.week, iso_tuple.day),
                             exp_iso)
```

I recognize that it adds a bit of complexity to the code, but it will give more granular output than we have now while also allowing us to easily add new examples if needed.

corona10 replied:
Great idea

pganssle reviewed (changes requested):
I haven't had time to do a full review yet, but I've added one comment about refactoring a test (feel free to dispute that if you disagree), and I do think that we should make the particular class used a private implementation detail and *not* expose it, per my comment [here](https://github.com/python/cpython/pull/15633/files#r319766276).

@taleinat I know that you asked for `IsoCalendarDate` to be documented - are you OK with us only guaranteeing that the return value has `year`, `week` and `day` fields but not exposing the actual class for `isinstance` or construction purposes? If so, I think we can just document those properties in the `isocalendar()` documentation.

corona10 committed:
bpo-24416: Apply code review

corona10 committed:
bpo-24416: Update docstrings

pganssle commented on Lib/datetime.py:
```diff
@@ -1091,7 +1116,7 @@ def isoweekday(self):
 
     def isocalendar(self):
-        """Return a 3-tuple containing ISO year, week number, and weekday.
+        """Return a IsoCalendarDate containing ISO year, week number, and weekday.
```
```suggestion
        """Return a named tuple containing ISO year, week number, and weekday.
```

pganssle commented on Lib/datetime.py:
```diff
@@ -8,6 +8,31 @@
 
+
+class IsoCalendarDate(tuple):
+    def __new__(cls, seq):
```
Is there a reason you chose this implementation rather than the other one? I suspect the other one will be somewhat more robust (even if the error message is different).

corona10 replied:
@pganssle This implementation allows the pickling exchange between python version and C version.

pganssle commented on Doc/whatsnew/3.9.rst:
```diff
@@ -116,6 +116,12 @@ Added the *indent* option to :func:`~ast.dump` which allows it to produce a
+datetime
+--------
+The ``isocalendar()`` methods of :class:`datetime.date` and
+:class:`datetime.datetime` now return a :class:`IsoCalendarDate`
```
```suggestion
:class:`datetime.datetime` now return a named tuple
```

Or

```suggestion
:class:`datetime.datetime` now return a :term:`named tuple`
```

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -1230,8 +1233,8 @@ Instance methods:
 
-   Return a 3-tuple, (ISO year, ISO week number, ISO weekday).  The same as
-   ``self.date().isocalendar()``.
+   Return a :class:`IsoCalendarDate`, (ISO year, ISO week number, ISO weekday).
```
```suggestion
   Return a named tuple, (ISO year, ISO week number, ISO weekday).
```

pganssle commented on Modules/_datetimemodule.c:
```diff
@@ -3383,7 +3420,7 @@ static PyMethodDef date_methods[] = {
 
     {"isocalendar", (PyCFunction)date_isocalendar,  METH_NOARGS,
-     PyDoc_STR("Return a 3-tuple containing ISO year, week number, and "
+     PyDoc_STR("Return a IsoCalendarDate containing ISO year, week number, and "
```
```suggestion
     PyDoc_STR("Return a named tuple containing ISO year, week number, and "
```

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -590,7 +590,7 @@ Instance methods:
 .. method:: date.isocalendar()
 
-   Return a 3-tuple, (ISO year, ISO week number, ISO weekday).
+   Return a :class:`IsoCalendarDate`, (ISO year, ISO week number, ISO weekday).
```
This link to `IsoCalendarDate` will or should be dead, as this is an undocumented class. I have made suggestions in many places to change it to "named tuple", in this documentation I suggest you reword it like this:

Return a :term:`named tuple` object with three components: ``year``, ``week`` and ``weekday``.

corona10 replied:
@pganssle 

> The pure-Python version defines a simple tuple sub-class with attribute access to the 
>  items, closely mimicking the structseq used by the C version. 

`:term:named tuple` is the right choice?
We are actually mimicking structseq aren't we?

tim-one replied:
Unfortunately, the Glossary has entries for both "named tuple" and "struct sequence", and it's clear as mud what the difference is intended to be.  At this point I'd be in favor of getting rid of the phrase "struct sequence" entirely everywhere in the docs, using only "named tuple" instead.  But that's outside the scope of this PR. 

The abbreviated name "structseq" doesn't mean anything to anyone outside CPython developers - it's an internal name used in CPython's C implementation code.

So, bottom line:  Paul's "`named tuple` object" looks best to me for now, although it will become more obviously clearer later, when the current "struct sequence"/"named tuple" confusion is purged from the docs.

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -606,6 +606,9 @@ Instance methods:
    4).isocalendar() == (2004, 1, 7)``.
 
+   .. versionchanged:: 3.9
+      Result changed from a tuple to a :class:`IsoCalendarDate`.
```
```suggestion
      Result changed from a tuple to a :term:`named tuple`.
```

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -606,6 +606,9 @@ Instance methods:
 
+   .. versionchanged:: 3.9
+      Result changed from a tuple to a :class:`IsoCalendarDate`.
+      It can't be unpickled in previous versions.
```
I think I would remove this line. In general pickling something in one version of Python and unpickling it in another version is not a supported operation and I think the fact that the class of the return value changed should be enough of a hint to people debugging this issue if it comes up.

taleinat commented on Lib/test/datetimetester.py:
```diff
@@ -1349,19 +1350,32 @@ def test_weekday(self):
+                self.assertEqual(d.isocalendar(), exp_iso)
+
+            # Check that the tuple contents are accessible by field name
+            with self.subTest(d=d, comparison="fields"):
```
The two loops here can be combined, which will also reduce the number of sub-tests.

pganssle replied:
@taleinat It's not clear to me what you mean by this - in my original proposal I deliberately used two loops because I thought it was more readable to 1. separate "test case creation" and "evaluating test cases" and 2. keep a lower level of nesting in each of the separate loops.

I also am not sure how combining the loops would reduce the number of sub-tests - without modifying precisely what you loop over the subtests are going to stay the same.

There's definitely a case to be made for not doing the full matrix of "every day of the week in each of these example weeks", but each pass of the loop is relatively inexpensive, and there's only 6 x 7 = 42 calls, tested twice. If we're concerned about it we could refactor like this, maybe:

```python
for d, exp_iso in test_cases:
    with self.subTest(d=d, subtest="call"):
        t = d.isocalendar()

    with self.subTest(d=d, subtest="tuple_comparison"):
        self.assertEqual(t, exp_iso)

    with self.subTest(d=d, subtest="field_comparison"):
        self.assertEqual((t.year, t.week, t.weekday), exp_iso)
```

Though `.isocalendar()` is fast, I suspect faster than entering and leaving a `subTest` context - it would only make sense to make this change to cleanly separate the three separate things we are actually testing here.

taleinat commented on Modules/_datetimemodule.c:
```diff
@@ -6516,6 +6553,17 @@ PyInit__datetime(void)
     PyModule_AddIntMacro(m, MAXYEAR);
 
+    /* IsoCalendarDate */
+    if (!initialized) {
```
In existing uses of structseq, I see both this initialization idiom used as well as checking whether `<type object>.tp_name == NULL`. The latter avoids using a module-level static variable, keeping the relevant logic all in one place. Would someone like to weigh in on which is preferred here?

tim-one replied:
> Would someone like to weigh in on which is preferred here?

Don't really care.  As a (very) old-timer here, I recognize the intent of the module-level static instantly, because that's how it was done from the start.  But the other way is fine too, provided there's a comment clarifying the intent of the otherwise-weird-looking test.

pganssle commented on Modules/_datetimemodule.c:
```diff
@@ -6516,6 +6553,17 @@ PyInit__datetime(void)
+        initialized = 1;
+    }
+    Py_INCREF((PyObject *) &StructIsoCalendarDateType);
+    PyModule_AddObject(m, "IsoCalendarDate", (PyObject *) &StructIsoCalendarDateType);
```
I think we can drop this line - there's no particular need to expose this as a public, constructable object in the C module. You'll also need to add `IsoCalendarDate` into [this list of cleaned up symbols](https://github.com/python/cpython/blob/3f4db4a0bab073b768fae958e93288bd5d24eadd/Lib/datetime.py#L2513) so it isn't still exposed.

pganssle commented on Lib/datetime.py:
```diff
@@ -8,6 +8,31 @@
 import sys
 
+
+class IsoCalendarDate(tuple):
```
I think the name of this symbol should be `_IsoCalendarDate`, because it turns out that `datetime` doesn't have an `__all__`, which would make `IsoCalendarDate` presumptively public. We want the repr to continue to say `IsoCalendarDate`, though, so I see three options:

1. Declare the class name as `IsoCalendarDate` then rename the symbol and delete the old symbol, like so:

```python
class IsoCalendarDate(tuple):
    ...

# The class name shouldn't start with an underscore, but the symbol should
_IsoCalendarDate = IsoCalendarDate
del IsoCalendarDate
```

2. Set `__name__` and `__qualname__`, like so:

```python
class _IsoCalendarDate(tuple):
    ...
_IsoCalendarDate.__name__ = "IsoCalendarDate"
_IsoCalendarDate.__qualname__ = "IsoCalendarDate"
```

3. Change the `__repr__` implementation to hard-code the class name, like so:

```python
    def __repr__(self):
        return ('datetime.IsoCalendarDate'
            f'(year={self[0]}, week={self[1]}, weekday={self[2]})')
```

There are positives and negatives from each approach and, ultimately, it doesn't really matter much. I think the first approach is the most natural.

pganssle replied:
Oh, I should also note, this is not a great place in the file for this class's definition. It's at the top, way before all the other class definitions. I would define it [right below the definition for `date`](https://github.com/python/cpython/blob/3f4db4a0bab073b768fae958e93288bd5d24eadd/Lib/datetime.py#L1140).

corona10 replied:
@pganssle 
This makes `python attribute lookup IsoCalendarDate on datetime failed` when pickling dump.

corona10 committed:
bpo-24416: Apply third's codereview

corona10 committed:
bpo-24416: Revert to support pickle

corona10 committed:
bpo-24416: Update misc/news

corona10 committed:
bpo-24416: Update __new__

pganssle commented on Lib/test/datetimetester.py:
```diff
@@ -1349,19 +1350,29 @@ def test_weekday(self):
+                test_cases.append((new_date, new_iso))
+
+        for d, exp_iso in test_cases:
+            with self.subTest(d=d, comparison="fields and tuple"):
```
I think these are at *least* two different tests, combining them into one subtest sorta defeats the point of subtest, can you revert to the old version, with two separate `subTest`s?

pganssle committed:
Change pickling behavior of isocalendar() objects

In order to leave IsocalendarDate as a private class and to improve what
backwards compatibility is offered for pickling the result of a
datetime.isocalendar() call, add a __reduce__ method to the named tuples
that reduces them to plain tuples.

Also add a test to this effect.

pganssle committed:
Rename IsoCalendarDate to _IsoCalendarDate

pganssle reviewed (changes requested):
I think the pickling interface is causing some problems, so I think the right thing to do at this point is to just pickle the underlying `tuple` object rather than trying to retain the `namedtuple`.

We can always remove that extension later (it won't be any more breaking than changing it now would be), and this allows us to keep `IsoCalendarDate` an implementation detail. I highly doubt that anyone will notice this, as it's very unlikely that anyone wants to pickle an object, pass it somewhere and then access its value by field (it's certainly possible, but part of the motivation for this change is that people are mostly just immediately accessing one or another of the fields of this tuple rather than passing around the result as a whole).

corona10 committed:
bpo-24416: Update datetime.rst

corona10 committed:
bpo-24416: Update unittest

pganssle committed:
Revert "Rename IsoCalendarDate to _IsoCalendarDate"

This reverts commit 448a2109dfde1558a082de78171d17580995a950.

pganssle committed:
Revert "Change pickling behavior of isocalendar() objects"

This reverts commit 104264121ea49795da87fc803a270bdda8fa21d2.

corona10 committed:
bpo-24416: Add pickle test

corona10 committed:
bpo-24416: Update C API to subclass tuple

corona10 committed:
bpo-24416: Change pickling behavior of isocalendar() objects

pganssle commented on Modules/_datetimemodule.c:
```diff
@@ -3225,6 +3227,174 @@ date_isoweekday(PyDateTime_Date *self, PyObject *Py_UNUSED(ignored))
+  ((year, week, weekday)\n\
+or via the object attributes as named in the above tuple.");
+
+typedef struct {
```
I have not spent enough time trying to implement this to know, but is there a reason why you've implemented a whole new class here rather than creating a `StructSeq` and then subclassing that rather than subclassing `tuple` directly?

Admittedly it's a bit neater to not complicate the MRO, but it does end up with a lot more implementation code when we really just want to override `__reduce__`.

corona10 replied:
@pganssle 
I actually intend to subclassing tuple directly with current implementation.
However, as I am not an expert of C API extension. 
if this is the wrong way, please let me know the standard way of subclassing tuple in C API.

other will work as same as tuple.

I am just overriding tuple with
`__repr__ `-> iso_calendar_date_repr
`__reduce__` -> isocalendardate_reduce
`__getset__` -> iso_calendar_date_methods
as same as pure python version.

tp_base is also PyTuple_Type,

corona10 replied:
You can see that IsoCalendarDate support basic tuple's operation without any define

pganssle replied:
Yes I noticed you are subclassing `tuple` directly, but that means you have to re-implement `StructSequence` yourself here. It's not a terribly complicated thing to implement, so it could be a reasonable thing to do for many reasons, I just wanted to check if you had deliberately ruled out using a StructSequence as the base instead of `tuple`.

corona10 replied:
@pganssle 
Oh, now I understood what you want to say (Sorry for my poor English)
As you know when we create this type with structsequence there was a test failure with test_os.
So I try to re-implement it as same as pure python and I notice that it works well.

> that means you have to re-implement StructSequence yourself here.

Yes, this is some kinds of complicated works but since we re-implement named tuple in pure python version. this works looks like same work around.

vstinner replied:
Can't reuse structseq type but override the type __reduce__ method?

corona10 replied:
@vstinner I am going to find a way to do this.
It does not look easy to me, if you can give a hint for me. It will be helpful :)

corona10 replied:
cc @pablogsal

vstinner replied:
If you write a PR using structseq, I suggest to create a separated PR. So we can compare both approaches. Create [WIP] PR without overriding `__reduce__()` if you want help.

corona10 replied:
Great idea

corona10 replied:
@vstinner @pablogsal 
https://github.com/python/cpython/pull/17651 is the experimental PR for comparing both approach.
I'd like to discuss how we will finalize this PR. (Do not change is the one of the options.)

pganssle commented on Modules/_datetimemodule.c:
```diff
@@ -3225,6 +3227,174 @@ date_isoweekday(PyDateTime_Date *self, PyObject *Py_UNUSED(ignored))
+[clinic start generated code]*/
+
+static PyObject *
+iso_calendar_date_new_impl(PyTypeObject *type, PyObject *seq)
```
If we stick with this version of the implementation, it's probably better to have a constructor that takes 3 C integers rather than a tuple of `PyLong_Long` objects, since the only place this should ever actually be constructed is in the code for `isocalendar()` and in the current form of this implementation, `isocalendar()` is first constructing an unnecessary intermediate `tuple` object, which is then broken apart to create the `IsoCalendarDate` object.

corona10 replied:
https://github.com/python/cpython/blob/0c989db57ec189511dc876973850bbd31c547ec0/Lib/datetime.py#L1216
So it means that we should update the pure python version of `__new__`?

corona10 committed:
bpo-24416: Update __new__

corona10 committed:
bpo-24416: Update whatsnew

vstinner commented on Lib/datetime.py:
```diff
@@ -1210,6 +1210,34 @@ def __reduce__(self):
+
+    def __repr__(self):
+        return (f'{self.__class__.__name__}'
+            f'(year={self[0]}, week={self[1]}, weekday={self[2]})')
```
nitpick: try to align it with the f on the first line ;-) put it on 3 lines if needed.

vstinner commented on Lib/datetime.py:
```diff
@@ -1210,6 +1210,34 @@ def __reduce__(self):
+        return self[2]
+
+    def __reduce__(self):
+        return (tuple, (tuple(self),))
```
Is it deliberate to deserialize it as a regular tuple? If yes, please add a comment to explain the rationale.

vstinner commented on Lib/test/datetimetester.py:
```diff
@@ -1355,19 +1356,45 @@ def test_weekday(self):
+        f.seek(0)
+        res = pickle.load(f)
+
+        self.assertEqual(res, (2019, 1, 2))
```
Please test also *res* type.

vstinner commented on Lib/test/datetimetester.py:
```diff
@@ -1355,19 +1356,45 @@ def test_weekday(self):
+        f = io.BytesIO()
+        pickle.dump(d.isocalendar(), f)
+        f.seek(0)
+        res = pickle.load(f)
```
You could just use dumps() and loads() to avoid a file, no?

corona10 committed:
bpo-24416: Apply vstinner's review

pppery commented on Lib/datetime.py:
```diff
@@ -1210,6 +1210,36 @@ def __reduce__(self):
 
+
+class IsoCalendarDate(tuple):
+
```
```suggestion

__slots__ = ()
```

corona10 replied:
@pppery Can you explain the reason about this suggestion?
